File: leetcode/883.三维形体投影面积.py
#
# @lc app=leetcode.cn id=883 lang=python3
#
# [883] 三维形体投影面积
#

# @lc code=start
from typing import List


class Solution:
    def projectionArea(self, grid: List[List[int]]) -> int:
        res = 0
        n = len(grid)
        max_y_z = [0] * n
        for x in range(n):
            max_z = 0
            for y in range(n):
                if grid[x][y] == 0:
                    continue
                res += 1
                max_z = max(max_z, grid[x][y])
                max_y_z[y] = max(max_y_z[y], grid[x][y])
            res += max_z
        res += sum(max_y_z)
        return res


# print(Solution().projectionArea([[1, 2], [3, 4]]))
# @lc code=end

File: leetcode/883.三维形体投影面积.go
package main

/*
 * @lc app=leetcode.cn id=883 lang=golang
 *
 * [883] 三维形体投影面积
 */

// @lc code=start
func Max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

func projectionArea(grid [][]int) int {
	res := 0
	n := len(grid)
	maxYZ := make([]int, n, n)
	for x := 0; x < n; x++ {
		maxZ := 0
		for y := 0; y < n; y++ {
			if grid[x][y] == 0 {
				continue
			}
			res += 1
			maxYZ[y] = Max(maxYZ[y], grid[x][y])
			maxZ = Max(maxZ, grid[x][y])
		}
		res += maxZ
	}
	for _, maxZ := range maxYZ {
		res += maxZ
	}
	return res
}

// func main() {
// 	fmt.Println(projectionArea([][]int{
// 		{1, 2},
// 		{3, 4},
// 	}))
// }

// @lc code=end
